File: insertion_merge/mergesort.py
def mergesort(arr: list, start: int, end: int):
  # Sorts in place

    if start >= end:
        return

    if end >= start + 2:
        mid = (start+end) // 2
        # print(f"mid is {mid}")
        mergesort(arr, start, mid)
        mergesort(arr, mid+1, end)

    merge(arr, start, end)


def merge(arr: list, start: int, end: int):
    if start >= end:
        return

    mid = start + (end-start) // 2
    start2 = mid + 1

    while start <= mid and start2 <= end:

        # Case 1: if first element is smaller than or equal to right element, i.e correct place
        if arr[start] <= arr[start2]:
            start += 1

        # Case 2: if first element is larger than right element, then shift the right element to before the left element
        else:
            # Right shift all elements from "start" up till element before "start2"
            value = arr[start2]
            for index in range(start2, start, -1):
                arr[index] = arr[index-1]
            # Copy over original start2 value into original start index
            arr[start] = value

            # Update all pointers for next iteration of while loop
            start += 1
            mid += 1
            start2 += 1


if __name__ == "__main__":
    arr = [50, 10, 2, 70, 3, 5]
    print(arr)

    mergesort(arr, 0, len(arr)-1)
    print(arr)

    # mergesort(arr)
    # print(arr)


# 5, 4, 3, 2, 1
# 3, 4, 5, 2, 1
# 3, 4, 5, 2, 1
# 2, 3, 4, 5, 1

File: insertion_merge/mergesort.c
#include <stdio.h>
#include <stdlib.h>

int compare(int, int);
void mergesort(int *, int, int);
void merge(int *, int, int);
void printlist(int *, int);

int main()
{
  int arr[] = {50, 10, 2, 70, 3, 5};
  // int arr[] = {9, 10, 11, 70, 3, 5};
  // int arr[] = {9, 8, 7, 6, 5, 4};
  // int arr[] = {4, 5, 6, 7, 8, 9};

  int length = 6;
  printlist(arr, length);
  mergesort(arr, 0, length - 1);
  printlist(arr, length);

  return 0;
}

void printlist(int *arr, int length)
{
  for (int i = 0; i < length; i++)
  {
    printf("%d ", arr[i]);
  }
  printf("\n");
}

// start and end are indexes
void mergesort(int *arr, int start, int end)
{
  if (start >= end)
  {
    return;
  }

  if (start <= end - 2)
  {
    int mid = start + (end - start) / 2;
    printf("mid is %d\n", mid);
    mergesort(arr, start, mid);
    mergesort(arr, mid + 1, end);
  }

  merge(arr, start, end);
}

void merge(int *arr, int start, int end)
{
  if (start >= end)
    return;

  // Merges arr from index n to index m
  int mid = start + (end - start) / 2;
  int a = start;
  int b = mid + 1;

  while (a <= mid && b <= end)
  {
    int cmp = compare(arr[a], arr[b]);

    // Case 1 : Right is greater than Left. Shift Right element to just before Left element.
    if (cmp > 0)
    {
      int tmp = arr[b]; // Save the value to be shifted
      for (int i = b; i > a; i--)
      {
        // Within the sorted portion, shift all values to the right. Leftmost value is unchanged.
        arr[i] = arr[i - 1];
      }
      arr[a] = tmp; // Save the tmp value into leftmost in sorted portion.

      // Increment all 3 pointers
      a++;
      b++;
      mid++;
    }
    else if (cmp < 0)
    {
      a++; // This value is in the right place. Just increment a by 1.
    }
    else
    {
      if (a == mid && b == end)
        break; // No movement needed if they are the last elements

      a++;              // Move a up by 1 because we want to insert the right side value only after this.
      int tmp = arr[b]; // Save the value to be shifted and move b forward by 1
      for (int i = b; i > a; i--)
      {
        // Within the sorted portion, shift all values to the right. Leftmost value is unchanged.
        arr[i] = arr[i - 1];
      }
      arr[a] = tmp;

      // Increment the 2 other pointers
      b++;
      mid++;
    }
  }
}

int compare(int val1, int val2)
{
  if (val1 > val2)
  {
    return 1;
  }
  else if (val1 == val2)
  {
    return 0;
  }
  else
  {
    return -1;
  }
}
